### PR TITLE
177 get tournament from db

### LINKED ISSUE #177: Add a way to retrieve tournaments from the Firebase

### PR DESCRIPTION
**Pull request by damiengenoud**
This PR adds:
- A function to retrieve a tournament from the database, using the tournament's ID.
- Utilitary functions to transform data snapshots of components of the tournament (like the Posts, LocalDateTime values, ...)

This PR does not contain functions to add posts or vote for a post, simply because I did not have the time to do it. This will have to be done during the last sprint.

Thanks in advance for your review!

closes #177

### CODE REVIEW

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt
Comment (Gafragni) [inline app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:1]: Well done, there are many tests implemented, covering many average cases and edge cases !

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt [hunk 1]
```diff
@@ -122,6 +125,63 @@ object FirebaseDatabaseUtils {
...
+        val month = getNumber(data?.child("monthValue"))?.toInt()
+        val dayOfMonth = getNumber(data?.child("dayOfMonth"))?.toInt()
+        val hour = getNumber(data?.child("hour"))?.toInt()
+        val minute = getNumber(data?.child("minute"))?.toInt()
+        val second = getNumber(data?.child("second"))?.toInt()
+        val nano = getNumber(data?.child("nano"))?.toInt()
+
+        if (year == null || month == null || dayOfMonth == null || hour == null || minute == null ||
+            second == null || nano == null
+        ) {
```
Comment (Gafragni) [inline app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:177]: Is there a nicer way to show this condition ?
This is purely about esthetics but that does not look right
```
        if (year == null || month == null || dayOfMonth == null || 
            hour == null || minute == null || second == null || nano == null
        ) {
```

**Review by Gafragni — APPROVED** [review]
Well done, functionalities are implemented, tested appropriately and test coverage is good.

I put some small remarks as suggestions but it is working and looks good for merging in my opinion.

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt [hunk 2]
```diff
@@ -263,6 +354,93 @@ class FirebaseDatabaseUtilsTest {
...
+        val transformedPosts = FirebaseDatabaseUtils.transformPostList(snapshot)
+
+        assertThat(transformedPosts.size, `is`(posts.size))
+    }
+
+    @Test
+    fun transformPostListReturnsEmptyListWithNullSnapshot() {
+        assertEquals(FirebaseDatabaseUtils.transformPostList(null), emptyList<TournamentPost>())
+    }
+
```
Comment (CristianSafta) [inline app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:443]: Suggestion: It would be nice to add a few more comments to improve code readability

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt [hunk 3]
```diff
@@ -363,4 +541,81 @@ class FirebaseDatabaseUtilsTest {
...
+    @Test
+    fun mapToTournamentReturnsNullWithNonExistingVisibilityValue() {
+        val mockVisibility = mockSnapshot("SomeVisibility")
+        val nonNullVisibilitySnapshot = mock(DataSnapshot::class.java)
+        `when`(nonNullVisibilitySnapshot.child("visibility"))
+            .thenReturn(mockVisibility)
+        assertEquals(FirebaseDatabaseUtils.mapToTournament(nonNullVisibilitySnapshot), null)
+    }
```
Comment (CristianSafta) [inline app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:620]: Suggestion: Same here

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt [hunk 4]
```diff
@@ -122,6 +125,63 @@ object FirebaseDatabaseUtils {
...
+    fun transformPostList(data: DataSnapshot?): List<TournamentPost> {
+        return data?.children?.mapNotNull {
+            transformPost(it)
+        } ?: emptyList()
+    }
+
+    /**
+     * Helper function to obtain a post from a database snapshot
+     * @param data the data snapshot containing the post
+     * @return the post, or null if an error occurred
```
Comment (CristianSafta) [inline app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:142]: Suggestion: Consider using an exception instead of returning null to indicate errors

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt [hunk 5]
```diff
@@ -122,6 +125,63 @@ object FirebaseDatabaseUtils {
...
+            return null
+        }
+
+        return TournamentPost(userId, run, votes, date, usersVotes.toMutableMap())
+    }
+
+    /**
+     * Helper function to obtain an object LocalDateTime from a data snapshot
+     * @param data the data snapshot containing the LocalDateTime
+     * @return the LocalDateTime, or null if an error occurred
```
Comment (CristianSafta) [inline app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:164]: Suggestion: Consider using an exception instead of returning null to indicate errors

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt [hunk 6]
```diff
@@ -207,4 +267,52 @@ object FirebaseDatabaseUtils {
             chatList = transformChatList(data.child(FirebaseKeys.USER_CHATS)),
         )
     }
+
+    /**
+     * Helper function to convert a data snapshot to a tournament
+     * @param data the data snapshot containing the tournament
+     * @return the tournament, or null if an error occurred
```
Comment (CristianSafta) [inline app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:274]: Suggestion: Consider using an exception instead of returning null to indicate errors

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt [hunk 7]
```diff
@@ -107,6 +109,95 @@ class FirebaseDatabaseUtilsTest {
...
+
+        val participants = mock(DataSnapshot::class.java)
+        val participantsSubNodes = tournament.participants.map {
+            val participant = mock(DataSnapshot::class.java)
+            `when`(participant.value).thenReturn(true)
+            `when`(participant.key).thenReturn(it)
+            participant
+        }
+        `when`(snapshot.child("participants")).thenReturn(participants)
+        `when`(participants.children).thenReturn(participantsSubNodes)
```
[the remark this answers is not in the document]
Comment (damiengenoud) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:190]: Participants is just a mock `DataSnapshot,` it does not contain any data, it basically just reacts to some calls (the ones in the `when` clauses). So here the only thing happenning with `participants` is to return a list of Snapshots when `partcipants.children` is called. 

It is implemented like that because for now in the real database, only `children` is called on `participants`, but if we need to use other properties like `child` or `key` we should add additionnal `when` clauses.

**Review by CristianSafta — APPROVED** [review]
The code is well-written and organized, uses asynchronous programming correctly, does exception handling, and makes effective use of Firebase's real-time database. The functions are clearly named, and each is accompanied by comprehensive comments detailing their purpose, parameters, and return types. This aids readability, understandability, and maintainability of the code. The proactive approach to error handling, through nullability checks and error conditions, is commendable. There are a few small suggestions of areas of improvement. However, overall the code is high quality and adheres to best software engineering practices.

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt [hunk 8]
```diff
@@ -263,6 +354,93 @@ class FirebaseDatabaseUtilsTest {
...
+            usersVotes = mutableMapOf("user1" to 1, "user2" to 1),
+        )
+        val postSnapshot = mockTournamentPost(post)
+        val transformedPost = FirebaseDatabaseUtils.transformPost(postSnapshot)
+        assertThat(transformedPost!!.userId, `is`(post.userId))
+        // Just to know if the run is the same one
+        assertThat(transformedPost.run.getCalories(), `is`(post.run.getCalories()))
+        assertThat(transformedPost.getVotes(), `is`(post.getVotes()))
+        assertThat(transformedPost.date, `is`(post.date))
+        assertThat(transformedPost.getUsersVotes(), `is`(post.getUsersVotes()))
```
[the remark this answers is not in the document]
Comment (damiengenoud) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:393]: Well for a the function `transformPost` to be correctly tests we need to check if it's components are the same, which is why I have put these assertions. Comparing the post directly does not work as the `Post` object is not the same, even though it has the same values.

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt (hunk 1 above)
Comment (damiengenoud) [reply app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:177]: I agree, I will change this with your format, thanks.

**Comment by damiengenoud** [comment]
@Gafragni 
Thanks for the review, I hope my answers to your questions are clear enough!

File: app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt (hunk 4 above)
Comment (damiengenoud) [reply app/src/main/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtils.kt:142]: Thanks for the suggestion, however I really don't think throwing an Exception in these cases is a better idea (but in nearly all cases I would completely agree). The reason is that it is much easier to handle "corrupted" data this way by using function like `mapNotNull` to handle lists of Post, Run, ... That way the corrupted data is simply ignored with a message in the logs, and we don't require endless try-catch blocks to prevent the app from crashing.

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt (hunk 3 above)
Comment (damiengenoud) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:620]: I can add comments in `mapToTournamentReturnsNullWithNonExistingVisibilityValue`, good suggestion! I think comments in the other tests would be really redundant though.

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt (hunk 2 above)
Comment (damiengenoud) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:443]: I really don't see where I could add comments that would improve the code readability here... maybe at line 431 ? It's a very simple mock so I don't really see the benefit.

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gafragni) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:190]: Oh i see now, thanks for the explaination !

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gafragni) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:425]: Oh I did not see the difference with the `Post` object, so it is necessary to put those tests here.
Well done then !

File: app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt [hunk 9]
```diff
@@ -263,6 +354,93 @@ class FirebaseDatabaseUtilsTest {
...
+                date = LocalDateTime.now().minusDays(3),
+                usersVotes = mutableMapOf("user1" to -1, "user2" to 1),
+            ),
+        )
+        val postsSnapshot = posts.map { mockTournamentPost(it) }
+        val snapshot = mock(DataSnapshot::class.java)
+        `when`(snapshot.children).thenReturn(postsSnapshot)
+        val transformedPosts = FirebaseDatabaseUtils.transformPostList(snapshot)
+
+        assertThat(transformedPosts.size, `is`(posts.size))
```
[the remark this answers is not in the document]
Comment (Gafragni) [reply app/src/androidTest/java/com/epfl/drawyourpath/database/FirebaseDatabaseUtilsTest.kt:468]: Yes, the previous explaination resolves this too, so nothing to say. It is all good.
